homeassistant/components/sleepiq/__init__.py

```diff
@@ -1,110 +1,227 @@
+        username=entry.data[CONF_USERNAME],
+    )
+
+    # Call the SleepIQ API to refresh data
+    await coordinator.async_config_entry_first_refresh()
+
+    # Listen to config entry updates
+    entry.async_on_unload(entry.add_update_listener(_async_signal_options_update))
+    entry.async_on_unload(
+        async_dispatcher_connect(
+            hass,
+            _config_entry_update_signal_name(entry),
+            coordinator.async_update_config,
+        )
+    )
+
+    hass.data[DATA_SLEEPIQ] = SleepIQHassData()
+    hass.data[DATA_SLEEPIQ].coordinators[entry.data[CONF_USERNAME]] = coordinator
+
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
 
     return True
 
 
-class SleepIQData:
-    """Get the latest data from SleepIQ."""
+async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Unload the config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(
+        config_entry, PLATFORMS
+    )
+
+    hass.data[DATA_SLEEPIQ].coordinators.pop(config_entry.data[CONF_USERNAME])
+
+    return unload_ok
 
-    def __init__(self, client):
-        """Initialize the data object."""
-        self._client = client
-        self.beds = {}
 
-        self.update()
+def _config_entry_update_signal_name(config_entry: ConfigEntry) -> str:
+    """Get signal name for updates to a config entry."""
+    return CONFIG_ENTRY_UPDATE_SIGNAL_TEMPLATE.format(config_entry.unique_id)
 
-    @Throttle(MIN_TIME_BETWEEN_UPDATES)
-    def update(self):
-        """Get the latest data from SleepIQ."""
-        self._client.login()
-        beds = self._client.beds_with_sleeper_status()
 
-        self.beds = {bed.bed_id: bed for bed in beds}
+async def _async_signal_options_update(
+    hass: HomeAssistant, config_entry: ConfigEntry
+) -> None:
+    """Signal config entry options update."""
+    async_dispatcher_send(
+        hass, _config_entry_update_signal_name(config_entry), config_entry
+    )
 
 
-class SleepIQSensor(Entity):
+class SleepIQDataUpdateCoordinator(DataUpdateCoordinator[dict[str, dict]]):
+    """SleepIQ data update coordinator."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        *,
+        client: Sleepyq,
+        update_interval: timedelta,
+        username: str,
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass, _LOGGER, name=f"{username}@SleepIQ", update_interval=update_interval
+        )
+        self.client = client
+
+    async def async_update_config(self, config_entry: ConfigEntry) -> None:
+        """Handle config update."""
+        self.update_interval = timedelta(
+            seconds=config_entry.options[CONF_SCAN_INTERVAL]
+        )
+
+    async def _async_update_data(self) -> dict[str, dict]:
+        data = {}
+        for bed in await self.hass.async_add_executor_job(
+            self.client.beds_with_sleeper_status
+        ):
```

Comment:
Make a function to do this that is called in the executor instead of jumping into the executor multiple times for each bed